libraries/lib-concurrency/ParallelReduce.h

```diff
@@ -0,0 +1,307 @@
+/**********************************************************************
+ 
+ Audacity: A Digital Audio Editor
+ 
+ @file ParallelReduce.h
+ 
+ Paul Licameli
+ 
+ **********************************************************************/
+#ifndef __AUDACITY_PARALLEL_REDUCE__
+#define __AUDACITY_PARALLEL_REDUCE__
+
+#include <cassert>
+#include <exception>
+#include <functional>
+#include <future>
+#include <iterator>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "ParallelProgress.h"
+#include "Splittable.h"
+
+namespace Parallel {
+
+//! Compute the result of a task that admits data-parallel decomposition of
+//! a domain described by one iterator range
+/*!
+ The simple subdivision assumes that the time spent in Task::run is proportional
+ to the size of the iterator range.
+
+ Copies of Task for non-empty, disjoint sub-ranges may be created and each of
+ them run on its own thread.  The calling thread always runs the original task
```

Comment:
Why? This limits the use of this class.

Reply:
At least when the supplied argument is 1, you know that the task just runs serially on the calling thread, so you can easily confirm first that the algorithm is serially correct before debugging it with multiple threads.

In the first application of this algorithm for tempo detection, the calling thread is the main thread and is the only one that may update the progress dialog.